File: examples/canonical-analytics/python/src/analytics_workbench/charts/funnel_chart.py
import plotly.graph_objects as go
from ..analytics.funnel import FunnelStages
from datetime import datetime

def build_funnel_figure(agg: FunnelStages) -> dict:
    fig = go.Figure()
    
    total_count = agg.counts[0] if agg.counts else 0
    
    if not agg.stages:
        fig.add_annotation(text="No data available", showarrow=False, font=dict(size=20))
        fig.update_layout(title_text="Session Funnel", xaxis_title_text="Count", yaxis_title_text="Stage")
    else:
        # Include drop-off percentage in hover text
        hover_texts = []
        for i in range(len(agg.stages)):
            pct = agg.drop_off_rates[i] * 100
            hover_texts.append(f"Drop-off: {pct:.1f}%")
            
        fig.add_trace(go.Funnel(
            y=agg.stages,
            x=agg.counts,
            name="Funnel",
            textinfo="value+percent initial",
            hovertext=hover_texts,
            hovertemplate="Stage: %{y}<br>Count: %{x}<br>%{hovertext}<extra></extra>"
        ))
        
        fig.update_layout(
            title_text="Session Funnel",
            xaxis_title_text="Count",
            yaxis_title_text="Stage"
        )
        
    fig_dict = fig.to_dict()
    fig_dict["meta"] = {
        "total_count": total_count,
        "filters_applied": [],
        "generated_at": datetime.utcnow().isoformat()
    }
    return fig_dict

File: examples/canonical-analytics/rust/src/charts/funnel_chart.rs
use super::to_figure_json;
use crate::analytics::FunnelStages;
use plotly::common::Title;
use plotly::{Bar, Plot};
use serde_json::Value;

pub fn build_funnel_figure(agg: &FunnelStages) -> Value {
    let mut plot = Plot::new();

    if agg.stages.is_empty() {
        let trace = Bar::new(Vec::<String>::new(), Vec::<i64>::new());
        plot.add_trace(trace);
        let layout = plotly::Layout::new().title(Title::with_text("No data for selected filters"));
        plot.set_layout(layout);
        return to_figure_json(&plot, 0, true);
    }

    // plotly.rs does not have a Funnel trace type in 0.10.0 natively easily accessible,
    // so the prompt says: "If plotly.rs does not have a Funnel type in the version you are using, 
    // use Bar with text annotations for drop-off. Document the substitution clearly."

    let total_count: i64 = *agg.counts.first().unwrap_or(&0);

    let mut text_annotations = Vec::new();
    for drop in &agg.drop_off_rates {
        text_annotations.push(format!("{:.1}% drop", drop * 100.0));
    }

    let trace = Bar::new(agg.stages.clone(), agg.counts.clone())
        .text_array(text_annotations)
        .hover_template("Stage: %{x}<br>Count: %{y}<extra></extra>");
    
    plot.add_trace(trace);

    let layout = plotly::Layout::new()
        .title(Title::with_text("Session Funnel"))
        .x_axis(plotly::layout::Axis::new().title(Title::with_text("Stage")))
        .y_axis(plotly::layout::Axis::new().title(Title::with_text("Count")));
    plot.set_layout(layout);

    to_figure_json(&plot, total_count, true)
}
